[#2545] Fix invalid text in exceptions raised for invalid input.

brunogola opened the pull request:
Problem
======

After parsing a SSH.COM key like the one below (and probably a bad OpenSSH key as well) `fromString_BLOB` was called without checking if BLOB is valid which resulted in a bad behavior as `fromString_BLOB` expects a proper blob.

```
---- BEGIN SSH2 PUBLIC KEY ----
Comment: "rsa-key-20141127"
Th5wpEcw1o9yIr1kTGsi7PbE3gphQEACMUmjXOdAVDU1ns1AgerxRzeh9GxfVe1Y
wgrPzfOgDBLYGTZlWf0ADPJd7f/2io9Lprq+fFd+TP/JdAsslg5GQy8zXq8ZNbKJ
hllW1u4lVGSUE/0pIURC+ptStP/2nsqfdVyhEJT8Il3aokRW5w==
---- END SSH2 PUBLIC KEY ----
```

Changes
=======

Added checks before calling `Key.fromString_BLOB()` to make sure that parsing the key went ok and to provide better error feedback whenever we fail to import an OpenSSH key or a SSH.com key.
Added a test to check this specific cases.
While working on it I did some changes to check if a SSH.com key has the END tag and changed tests to test this as well.

adiroiban commented on chevah/keycert/ssh.py:
```diff
@@ -299,7 +299,7 @@ def fromString(cls, data, type=None, passphrase=None):
             # we consider it too short.
             raise BadKeyError('Key is too short.')
-        except (struct.error,  binascii.Error, TypeError):
+        except (struct.error,  binascii.Error, TypeError, AssertionError):
```
why are we catching an assertion error here ... together with all other errors ?

if assertionerror should be handled, the it should be raised as something else.

brunogola replied:
all this errors, including the assertion, means that the parsing went wrong. 

in my point of view fromString_SSHCOM_PUBLIC... should not be calling fromString_BLOB the way it is. 
I say this because there is a verification when you call fromString() passing a blob, it calls _guessStringType and there it will only call fromString_BLOB if the blob starts with '\x00\x00\x00\x07ssh-'.. 

so I could check if the blob is valid before trying to extract the type (inside fromString_BLOB) or validate before calling it. i opted for the second option because this way I could keep the error messages consistent when using fromString() passing a OpenSSH or a SSH.COM key. otherwise the error message when trying to parse this kind of keys would be related to the blob, which, in my opinion, could lead to confusion.

still i don't think this code deals in a good way with all the possible errors when parsing keys, but changing it would be a way bigger task then the one I wanted to solve here.  considering that, I think that catching the assertionError here makes sense, as it will lead to raising BadKeyError with a consistent message..

adiroiban replied:
It is ok to raise a non-BadKeyError, just don't use the AssertionError for this :)

we are using AssertionError for 'imposibile' error conditions and for preventing using an API in in the wrong way.

we don't use AssertionError for input validation.

adiroiban replied:
I was hoping that since we touch this part, we can dig deeper to find the root cause of this type of errors and try to fix it in a generic way so that we can have a nice keycert API which always return BadKeyError exception with valid text messages

adiroiban commented on chevah/keycert/tests/test_ssh.py:
```diff
@@ -1208,6 +1207,13 @@ def test_fromString_PUBLIC_SSHCOM_RSA_invalid_payload(self):
 
         self.assertKeyParseError(content)
 
+        # This will parse without errors but the blob is invalid
```
this needs a separate test, with a dedicated docstring describing the targeted test scenario

adiroiban commented on chevah/keycert/tests/test_ssh.py:
```diff
@@ -864,8 +864,7 @@ def test_fromBlob(self):
         self.assertEqual(dsaKey.data(), {'p': 2L, 'q': 3L, 'g': 4L, 'y': 5L})
-        self.assertRaises(
-            keys.BadKeyError, keys.Key.fromString, badBlob)
+        self.assertBadKey(badBlob, "Unknown blob type: ssh-bad.")
```
for this test, maybe we should use a key type with an invalid utf-8

adiroiban commented on chevah/keycert/ssh.py:
```diff
@@ -676,7 +676,7 @@ def _fromString_PRIVATE_BLOB(cls, blob):
             return dsakey
         else:
-            raise BadKeyError('unknown blob type: %s' % keyType)
+            raise BadKeyError('Unknown blob type: %s.' % keyType)
```
I think that here we should use %r to make sure we don't try to decode random data.... and this should be valid for any other exception in which we inject a non-sanitized input

adiroiban commented on chevah/keycert/ssh.py:
```diff
@@ -690,6 +690,7 @@ def _fromString_PUBLIC_OPENSSH(cls, data):
         @raises BadKeyError: if the blob type is unknown.
         """
         blob = base64.decodestring(data.split()[1])
+        assert cls._guessStringType(blob) == 'blob'
```
just raise BadKeyError('Helpful unique message') instead of this generic assert.

we don't use assertion in this way as I consider it bad practice.

but I don't know if this is the right way to fix it.

_guessStringType is just a soft check... you might have a data of type blob but with other type of encoding errors.

We should update _fromString_BLOB and have it raise BadKeyError on errors

your implementation breaks the docstring as it will not raise AssertionError instead of BadKeyError

adiroiban commented:
Many thanks for working on this. 

I am happy to see more work on this project.

While the project is old and the code quality is not perfect I hope that we can work on improving it to make it a great and robust lib for reading/writing/generating/converting ssh keys and x509 related files

-------

I don't like the current fix as it just fixed a symptom, rather than trying to see the root cause of the error and fixing it.

I would say that the main problem here is that we treat invalid input as 'valid text' instead of 'invalid text'.

From my point of view a fix would be to use `%r` + truncate for all input values which are included in the text of an exception.

You can have a key with a type which is 100kb long , and including the whole type in the error message would not help to much

------

For the test side I was expecting to see some new tests with input data which is not valid text

Also, the new test changes are false positives as they don't cover any new use case.

For example try to reverse your main changes and run the new tests unchanges. If they tests are valid, they should fail... but I have reverted your changes and the test still pass

```
diff --git a/chevah/keycert/ssh.py b/chevah/keycert/ssh.py
index 5f57306..1c40973 100644
--- a/chevah/keycert/ssh.py
+++ b/chevah/keycert/ssh.py
@@ -299,7 +299,7 @@ class Key(object):
             # Most probably some parts are missing from the key, so
             # we consider it too short.
             raise BadKeyError('Key is too short.')
-        except (struct.error,  binascii.Error, TypeError, AssertionError):
+        except (struct.error,  binascii.Error, TypeError):
             raise BadKeyError('Fail to parse key content.')
 
     def toString(self, type, extra=None):
@@ -690,7 +690,6 @@ class Key(object):
         @raises BadKeyError: if the blob type is unknown.
         """
         blob = base64.decodestring(data.split()[1])
-        assert cls._guessStringType(blob) == 'blob'
         return cls._fromString_BLOB(blob)
 
     @classmethod
@@ -1075,7 +1074,6 @@ class Key(object):
         @raises BadKeyError: if the blob type is unknown.
         """
         blob = cls._getSSHCOMKeyContent(data)
-        assert cls._guessStringType(blob) == 'blob'
         return cls._fromString_BLOB(blob)
 
     @classmethod
```

When fixing bugs it is very important to write the test first, see it fail and then proceed with fixing the code.

needs-changes

adiroiban commented:
I have also updated the title of this review..

It was missing the Trac ticket id.

The title should be about *what* is done/fixed , not *how* is done/fixed. We can get how from the diff :)

brunogola commented:
I did some new changes, removing the assertions and fixing one of my tests.

Now if you try to run test_fromString_PUBLIC_OPENSSH_invalid_blob or test_fromString_PUBLIC_SSHCOM_RSA_invalid_blob in master you can see the problem

With the other changes I was trying to get more consistent error messages, as getting 'Unknown blob type' when trying to parse a SSH.com or OpenSSH key, for me, is misleading. I think that getting 'Fail to parse OpenSSH key content' or the more specific 'Fail to find END tag for SSH.com key.' might be more useful.

Also to avoid dealing with non-ascii keyTypes when parsing a blob I added a verification, then we can have a specific error message. Not sure if it is that useful but I think it is better than printing non-ascii data.

I still think there is how to improve the key validation, type guessing, etc, in this code, but my priority was to fix the server.

brunogola commented:
needs-review

adiroiban commented on chevah/keycert/tests/test_ssh.py:
```diff
@@ -859,13 +859,14 @@ def test_fromBlob(self):
-        self.assertRaises(
-            keys.BadKeyError, keys.Key.fromString, badBlob)
+        self.assertBadKey(badBlob, 'Invalid non-ascii blob type.')
+        self.assertBadKey(badKey, 'Unknown blob type: \'ssh-bad\'')
```
here we have 2 tests. Please break this into 2 tests, each with its own description of the test scenario.

I know that the current tests is not great, but new code should be written based on current best practices

adiroiban commented on chevah/keycert/ssh.py:
```diff
@@ -629,6 +629,13 @@ def _fromString_BLOB(cls, blob):
+            # keyType is supposed to contain only ascii characters
+            keyType.decode('ascii')
+        except UnicodeDecodeError:
+            raise BadKeyError('Invalid non-ascii blob type.')
```
For such errors please also include the truncated repr of 'keyType' in the error message.

It will not help the end user, but it will help the development/support team when an user is reporting an error.

The user report an error like `Invalid non-ascii blob type.` but then we will have no idea if this is a bug in our software or user provided invalid input.

If user will report something like `Invalid non-ascii blob type. bla-bla` we can get an idea of what was going on there.

---------

I am not sure if this is required. I prefer to leave this out and have this case handled by `BadKeyError('Unknown blob type: %r' % keyType[:30])` 

As a general design rule, we should just use EAFF and make sure the error messages are sane.

------

While keyType are expected to be ascii, in the RFC i have not found any restriction for that... so they should be treated as bytes.

We should update chevah-keycert and port it to Py3 (already done in Twisted) and after that it will be clear which value is text and which is bytes.

brunogola replied:
I got confused by reading the RFC https://tools.ietf.org/html/rfc4251#section-5:

"Strings are also used to store text.  In that case, US-ASCII is used for internal names, and ISO-10646 UTF-8 for text that might be displayed to the user." 

But you are right, strings can also contain binary data according to that RFC.

adiroiban commented on chevah/keycert/ssh.py:
```diff
@@ -676,7 +683,7 @@ def _fromString_PRIVATE_BLOB(cls, blob):
             return dsakey
         else:
-            raise BadKeyError('unknown blob type: %s' % keyType)
+            raise BadKeyError('Unknown blob type: %r' % keyType)
```
why not also truncate here ?

adiroiban commented on chevah/keycert/ssh.py:
```diff
@@ -690,6 +697,8 @@ def _fromString_PUBLIC_OPENSSH(cls, data):
         """
         blob = base64.decodestring(data.split()[1])
+        if cls._guessStringType(blob) != 'blob':
+            raise BadKeyError("Fail to parse OpenSSH key content.")
```
You will get a nice 'OpenSSH key' error only for this particular case.

what to you say about having something like this?
```
try:
    return cls._fromString_BLOB(blob)
except BadKeyError as error:
    raise BadKeyError('OpenSSH key error. %s' % (error.message'))
```

adiroiban replied:
If required, you can move the `cls._guessStringType(blob) != 'blob'` check in `_fromString_BLOB`.... but I hope that we can design the code with as few checks and if conditionals as possible

adiroiban replied:
There were a couple of good blog post on the internet describing why to many if statements are bad practice, but I can not find them right now.... but basically, less is more... and if you can get the same/similar with less if statements, that solution is better :)

adiroiban replied:
We can also refactor this as


```
    return cls._fromString_BLOB(blob, key_name='OpenSSH')
```

and _fromString_BLOB will include the key_name for all the raised messages

adiroiban commented:
Thanks for the changes.

I think that for this bug you forgot to ask the 5 why :)

From my point of view the root problem is that we treat all input values are text, rather than bytes.

So to fix the root cause, we should refact the whole keycert code to use `%r` when we inject unfiltered input data.

You have fixed the issue for ssh.com and openssh but if later we add another blob based key format, we might forget to add all the custom/duplicate input validation and we might end up again with similar issues. ... and without friendly messages.

The idea is that when you found an error, try to dig deeper before starting to fix it at the level at which the error was initially found.

--------

To fix this, I think that is easier to first port it to py3 and then come back with another ticket and fix this issue.

so for this branch is ok if you just implement the `%r` and truncation and maybe `cls._fromString_BLOB(blob, key_name='OpenSSH')`

--------

We can look at providing user friendly messages in a separate ticket after the py3 port.



needs-changes

brunogola commented:
needs-review

adiroiban commented:
Changes looks good.

I think that since we are here, we should review the code and fix all other usage of `%s` instead of `%r`

for example

(12:14:55) adiroiban: ex https://github.com/chevah/chevah-keycert/blob/2545-fix-parsing-openssh-keys/chevah/keycert/ssh.py#L291
(12:15:05) adiroiban: https://github.com/chevah/chevah-keycert/blob/2545-fix-parsing-openssh-keys/chevah/keycert/ssh.py#L325
(12:16:06) adiroiban: https://github.com/chevah/chevah-keycert/blob/2545-fix-parsing-openssh-keys/chevah/keycert/ssh.py#L865
(12:16:19) adiroiban: https://github.com/chevah/chevah-keycert/blob/2545-fix-parsing-openssh-keys/chevah/keycert/ssh.py#L987
(12:16:41) adiroiban: https://github.com/chevah/chevah-keycert/blob/2545-fix-parsing-openssh-keys/chevah/keycert/ssh.py#L1136
(12:16:52) adiroiban: https://github.com/chevah/chevah-keycert/blob/2545-fix-parsing-openssh-keys/chevah/keycert/ssh.py#L1146
(12:17:01) adiroiban: https://github.com/chevah/chevah-keycert/blob/2545-fix-parsing-openssh-keys/chevah/keycert/ssh.py#L1153
(12:17:23) adiroiban: https://github.com/chevah/chevah-keycert/blob/2545-fix-parsing-openssh-keys/chevah/keycert/ssh.py#L1396
(12:17:34) adiroiban: https://github.com/chevah/chevah-keycert/blob/2545-fix-parsing-openssh-keys/chevah/keycert/ssh.py#L1416
(12:18:22) adiroiban: https://github.com/chevah/chevah-keycert/blob/2545-fix-parsing-openssh-keys/chevah/keycert/ssh.py#L1457 for private_mac
(12:18:43) adiroiban: there are the places which I think might have the same issue


needs-changes

brunogola commented:
needs-review

adiroiban commented on chevah/keycert/ssh.py:
```diff
@@ -1183,19 +1183,19 @@ def _fromString_PRIVATE_SSHCOM(cls, data, passphrase):
 
         encryption_key = None
-        if cipher_type.lower() == '3des-cbc':
+        if cipher_type.lower() == b'3des-cbc':
```
is a bit strange that we call 'lower' on a set of bytes... but I can see that py3  still has the lower method for bytes.

adiroiban commented on chevah/keycert/common.py:
```diff
@@ -5,11 +5,29 @@
+import sys
+
+_PY3 = sys.version_info > (3,)
+if _PY3:
```
You can add the no-cover praga marker 

http://coverage.readthedocs.org/en/coverage-4.0.3/excluding.html

adiroiban commented on chevah/keycert/ssh.py:
```diff
@@ -21,21 +24,26 @@
+if not _PY3:
+    izip = itertools.izip
+else:
+    izip = zip
```
why not move it in keycert.common together with the other py3 conditionals ?

adiroiban commented on chevah/keycert/ssh.py:
```diff
@@ -237,9 +243,14 @@ def __repr__(self):
+        if _PY3 and isinstance(type, bytes):
+            type = type.decode('ascii')
+        elif not _PY3 and isinstance(type, unicode):
+            type = type.encode('ascii')
```
this is very ugly... I would argue that we should fix self.type() so that it will always return the same data type and that is bytes.

adiroiban commented:
Looks good.

Please try to group as many py3 specific code into a single block and instruct coverage.py to ignore that part.

Thanks!

changes-approved

brunogola commented:
thanks for the review. 

I addressed your suggestions but regarding type() still think it should not return bytes, so I followed twisted approach in a better and added a native_string helper. what do you think?

needs_review

adiroiban commented:
looks good. thanks

changes-approved

brunogola commented:
merged in a52685c50f0ced93667464403576edadd4a74559
